ember-file-upload/src/system/data-transfer-wrapper.ts

```diff
@@ -43,6 +47,81 @@ export default class DataTransferWrapper {
     return this.files.length ? this.files : this.items;
   }
 
+  async getFilesAndDirectories() {
+    const files: File[] = [];
+
+    const readEntry = async (entry: FileSystemEntry): Promise<File> => {
+      return new Promise((resolve, reject) => {
+        if (entry.isFile) {
+          (entry as FileSystemFileEntry).file((fileEntry: File) => {
+            resolve(fileEntry);
+          });
+        } else {
+          reject('Directory contains nested directories');
+        }
+      });
+    };
+
+    const getEntry = (
+      item: FutureProofDataTransferItem,
+    ): FileSystemDirectoryEntry => {
+      // In the future this method name might change, so already implementing it like this if needed
+      // https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry
+      if (typeof item.getAsEntry === 'function') {
+        return item.getAsEntry();
+      }
+
+      return item.webkitGetAsEntry() as FileSystemDirectoryEntry;
+    };
+
+    const readAllFilesInDirectory = (item: DataTransferItem): Promise<File[]> =>
+      new Promise((resolve) => {
+        getEntry(item)
+          ?.createReader()
+          ?.readEntries(async (entries: FileSystemEntry[]) => {
+            const readFiles: File[] = await Promise.all(
+              entries.map(readEntry),
+            ).catch((err) => {
+              throw err;
+            });
+            resolve(readFiles.filter((file) => file !== undefined) as File[]);
+          });
+      });
+
+    const readDataTransferItem = async (
+      item: DataTransferItem,
+    ): Promise<File[]> => {
+      if (getEntry(item)?.isDirectory) {
+        const directoryFile = item.getAsFile() as File;
+        const filesInDirectory: File[] = await readAllFilesInDirectory(item);
+        return [directoryFile, ...filesInDirectory];
+      } else {
+        const fileItem = item.getAsFile() as File;
+        return [fileItem];
+      }
+    };
+
+    if (this.dataTransfer?.items) {
+      const allFilesInDataTransferItems: File[][] = await Promise.all(
+        Array.from(this.dataTransfer?.items).map(readDataTransferItem),
+      );
+
+      const flattenedFileArray: File[] = allFilesInDataTransferItems.reduce(
+        (flattenedList, fileList) => {
+          return [...flattenedList, ...fileList];
+        },
+        [],
+      );
+
+      files.push(...flattenedFileArray);
+    } else {
+      const droppedFiles: File[] = Array.from(this.dataTransfer?.files ?? []);
+      files.push(...droppedFiles);
+    }
+
+    return files;
```

Comment:
Not sure we need to define a files variable and mutate it. Each of these branches could just be an outright return

Reply:
Made it immediate returns